bindings/wasm/output.go

```diff
@@ -0,0 +1,192 @@
+	guest []byte
+}
+
+type outputBinding struct {
+	logger        logger.Logger
+	runtimeConfig wazero.RuntimeConfig
+	moduleConfig  wazero.ModuleConfig
+
+	runtime wazero.Runtime
+	module  wazero.CompiledModule
+
+	instanceCounter uint64
+}
+
+var (
+	_ bindings.OutputBinding = (*outputBinding)(nil)
+	_ io.Closer              = (*outputBinding)(nil)
+)
+
+func NewWasmOutput(logger logger.Logger) bindings.OutputBinding {
+	return &outputBinding{
+		logger: logger,
+
+		// The below ensures context cancels in-flight wasm functions.
+		runtimeConfig: wazero.NewRuntimeConfig(),
+
+		// The below violate sand-boxing, but allow code to behave as expecteout.
+		moduleConfig: wazero.NewModuleConfig().
+			WithRandSource(rand.Reader).
+			WithSysNanosleep().
+			WithSysWalltime().
+			WithSysNanosleep(),
+	}
+}
+
+func (out *outputBinding) Init(ctx context.Context, metadata bindings.Metadata) (err error) {
+	meta, err := out.getInitMetadata(metadata)
+	if err != nil {
+		return fmt.Errorf("wasm: failed to parse metadata: %w", err)
+	}
+
+	out.runtime = wazero.NewRuntimeWithConfig(ctx, out.runtimeConfig)
+	out.module, err = out.runtime.CompileModule(ctx, meta.guest)
+	if err != nil {
+		_ = out.runtime.Close(ctx)
+		return fmt.Errorf("wasm: error compiling binary: %w", err)
+	}
+
+	switch detectImports(out.module.ImportedFunctions()) {
+	case modeWasiP1:
+		_, err = wasi_snapshot_preview1.Instantiate(ctx, out.runtime)
+	}
+
+	if err != nil {
+		return fmt.Errorf("wasm: error instantiating host functions: %w", err)
+	}
+	return
+}
+
+func (out *outputBinding) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// Currently, concurrent modules can conflict on name. Make sure we have
+	// a unique one.
+	moduleName := strconv.FormatUint(atomic.AddUint64(&out.instanceCounter, 1), 10)
+	moduleConfig := out.moduleConfig.WithName(moduleName)
+
+	if len(req.Data) > 0 {
+		moduleConfig = moduleConfig.WithStdin(bytes.NewReader(req.Data))
+	}
+
+	var stdout bytes.Buffer
+	moduleConfig = moduleConfig.WithStdout(&stdout)
+
+	if args, ok := req.Metadata["args"]; ok {
+		argsSlice := strings.Split(args, ",")
+		// prepend arg0 (program name) unless we want users to supply it.
+		argsSlice = append([]string{out.module.Name()}, argsSlice...)
+		moduleConfig = moduleConfig.WithArgs(argsSlice...)
+	}
+
+	// Instantiating executes the guest's main function (exported as _start).
+	_, err := out.runtime.InstantiateModule(ctx, out.module, moduleConfig)
+
+	// Return stdout if there was no error.
+	if err == nil {
+		return &bindings.InvokeResponse{Data: stdout.Bytes()}, nil
+	}
+
+	// Handle any error, noting exit zero is not one.
+	if exitErr, ok := err.(*sys.ExitError); ok && exitErr.ExitCode() == 0 {
+		return &bindings.InvokeResponse{Data: stdout.Bytes()}, nil
+	}
+	return nil, err
+}
+
+func (out *outputBinding) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{bindings.GetOperation}
```

Comment:
Is this really a `Get` operation? we can support any operation name that matches what is being done as closely as possible - perhaps `Execute`?